File: tdf-hq/src/TDF/API/CommerceOperations.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators #-}

module TDF.API.CommerceOperations
  ( CommerceOperationsAPI
  , CommercePaymentOverviewDTO(..)
  , CommerceProviderAccountDTO(..)
  , CommerceProviderCapabilityDTO(..)
  , CommercePaymentIntentSummaryDTO(..)
  , CommerceAmountComponentSummaryDTO(..)
  , CommerceCommissionSummaryDTO(..)
  , CommerceRefundSummaryDTO(..)
  , CommerceDisputeSummaryDTO(..)
  , CommerceReconciliationSummaryDTO(..)
  , CommerceSettlementSummaryDTO(..)
  , CommerceSellerBalanceSummaryDTO(..)
  , CommercePayoutSummaryDTO(..)
  , CommerceProviderEventDTO(..)
  , CommerceProviderEventReplayCreate(..)
  ) where

import           Data.Aeson (FromJSON(..), ToJSON, genericParseJSON)
import           Data.Int (Int64)
import           Data.Text (Text)
import           Data.Time (UTCTime)
import           GHC.Generics (Generic)
import           Servant

import           TDF.API.Types (strictObjectOptions)

-- | Redacted provider readiness. Merchant account references, feature-flag
-- keys, secrets and capability source payloads are intentionally absent.
data CommerceProviderAccountDTO = CommerceProviderAccountDTO
  { cpaProvider         :: Text
  , cpaEnvironment      :: Text
  , cpaStatus           :: Text
  , cpaContractStatus   :: Text
  , cpaCredentialStatus :: Text
  , cpaSettlementCurrency :: Text
  , cpaEnabled          :: Bool
  , cpaFeatureEnabled   :: Bool
  , cpaVerifiedAt       :: Maybe UTCTime
  , cpaDisabledReason   :: Maybe Text
  , cpaCapabilities     :: [CommerceProviderCapabilityDTO]
  } deriving (Eq, Show, Generic)

instance ToJSON CommerceProviderAccountDTO
instance FromJSON CommerceProviderAccountDTO

data CommerceProviderCapabilityDTO = CommerceProviderCapabilityDTO
  { cpcPaymentMethod      :: Text
  , cpcCapability         :: Text
  , cpcVerificationStatus :: Text
  , cpcVerifiedAt         :: Maybe UTCTime
  } deriving (Eq, Show, Generic)

instance ToJSON CommerceProviderCapabilityDTO
instance FromJSON CommerceProviderCapabilityDTO

data CommercePaymentIntentSummaryDTO = CommercePaymentIntentSummaryDTO
  { cpiStatus          :: Text
  , cpiCurrency        :: Text
  , cpiCount           :: Int64
  , cpiAmountMinor     :: Int64
  , cpiAuthorizedMinor :: Int64
  , cpiCapturedMinor   :: Int64
  , cpiRefundedMinor   :: Int64
  } deriving (Eq, Show, Generic)

instance ToJSON CommercePaymentIntentSummaryDTO
instance FromJSON CommercePaymentIntentSummaryDTO

data CommerceAmountComponentSummaryDTO = CommerceAmountComponentSummaryDTO
  { cacComponentType :: Text
  , cacSource        :: Text
  , cacCurrency      :: Text
  , cacCount         :: Int64
  , cacAmountMinor   :: Int64
  } deriving (Eq, Show, Generic)

instance ToJSON CommerceAmountComponentSummaryDTO
instance FromJSON CommerceAmountComponentSummaryDTO

data CommerceCommissionSummaryDTO = CommerceCommissionSummaryDTO
  { ccmProvider         :: Text
  , ccmEnvironment      :: Text
  , ccmCurrency         :: Text
  , ccmCount            :: Int64
  , ccmBasisAmountMinor :: Int64
  , ccmCommissionMinor  :: Int64
  , ccmProviderFeeMinor :: Int64
  , ccmTaxMinor         :: Int64
  , ccmSellerNetMinor   :: Int64
  } deriving (Eq, Show, Generic)

instance ToJSON CommerceCommissionSummaryDTO
instance FromJSON CommerceCommissionSummaryDTO

data CommerceRefundSummaryDTO = CommerceRefundSummaryDTO
  { crfProvider    :: Text
  , crfEnvironment :: Text
  , crfStatus      :: Text
  , crfCurrency    :: Text
  , crfCount       :: Int64
  , crfAmountMinor :: Int64
  } deriving (Eq, Show, Generic)

instance ToJSON CommerceRefundSummaryDTO
instance FromJSON CommerceRefundSummaryDTO

data CommerceDisputeSummaryDTO = CommerceDisputeSummaryDTO
  { cdsProvider    :: Text
  , cdsEnvironment :: Text
  , cdsKind        :: Text
  , cdsStatus      :: Text
  , cdsCurrency    :: Text
  , cdsCount       :: Int64
  , cdsAmountMinor :: Int64
  } deriving (Eq, Show, Generic)

instance ToJSON CommerceDisputeSummaryDTO
instance FromJSON CommerceDisputeSummaryDTO

data CommerceReconciliationSummaryDTO = CommerceReconciliationSummaryDTO
  { crsProvider      :: Text
  , crsEnvironment   :: Text
  , crsStatus        :: Text
  , crsCurrency      :: Maybe Text
  , crsCount         :: Int64
  , crsExpectedMinor :: Int64
  , crsActualMinor   :: Int64
  } deriving (Eq, Show, Generic)

instance ToJSON CommerceReconciliationSummaryDTO
instance FromJSON CommerceReconciliationSummaryDTO

data CommerceSettlementSummaryDTO = CommerceSettlementSummaryDTO
  { cssProvider         :: Text
  , cssEnvironment      :: Text
  , cssStatus           :: Text
  , cssCurrency         :: Text
  , cssCount            :: Int64
  , cssGrossMinor       :: Int64
  , cssFeeMinor         :: Int64
  , cssWithholdingMinor :: Int64
  , cssRefundMinor      :: Int64
  , cssChargebackMinor  :: Int64
  , cssNetMinor         :: Int64
  } deriving (Eq, Show, Generic)

instance ToJSON CommerceSettlementSummaryDTO
instance FromJSON CommerceSettlementSummaryDTO

data CommerceSellerBalanceSummaryDTO = CommerceSellerBalanceSummaryDTO
  { csbProvider       :: Text
  , csbEnvironment    :: Text
  , csbAvailability   :: Text
  , csbCurrency       :: Text
  , csbEntryCount     :: Int64
  , csbNetAmountMinor :: Int64
  } deriving (Eq, Show, Generic)

instance ToJSON CommerceSellerBalanceSummaryDTO
instance FromJSON CommerceSellerBalanceSummaryDTO

data CommercePayoutSummaryDTO = CommercePayoutSummaryDTO
  { cpsProvider    :: Text
  , cpsEnvironment :: Text
  , cpsStatus      :: Text
  , cpsCurrency    :: Text
  , cpsCount       :: Int64
  , cpsAmountMinor :: Int64
  } deriving (Eq, Show, Generic)

instance ToJSON CommercePayoutSummaryDTO
instance FromJSON CommercePayoutSummaryDTO

data CommercePaymentOverviewDTO = CommercePaymentOverviewDTO
  { cpoGeneratedAt              :: UTCTime
  , cpoProviderAccounts         :: [CommerceProviderAccountDTO]
  , cpoPaymentIntents           :: [CommercePaymentIntentSummaryDTO]
  , cpoAmountComponents         :: [CommerceAmountComponentSummaryDTO]
  , cpoCommissions              :: [CommerceCommissionSummaryDTO]
  , cpoRefunds                  :: [CommerceRefundSummaryDTO]
  , cpoDisputes                 :: [CommerceDisputeSummaryDTO]
  , cpoReconciliationExceptions :: [CommerceReconciliationSummaryDTO]
  , cpoSettlements              :: [CommerceSettlementSummaryDTO]
  , cpoSellerBalances           :: [CommerceSellerBalanceSummaryDTO]
  , cpoPayouts                  :: [CommercePayoutSummaryDTO]
  } deriving (Eq, Show, Generic)

instance ToJSON CommercePaymentOverviewDTO
instance FromJSON CommercePaymentOverviewDTO

-- | Sensitive fields are intentionally excluded from this operator DTO. In
-- particular, the encrypted provider payload and merchant account reference
-- never leave the backend.
data CommerceProviderEventDTO = CommerceProviderEventDTO
  { cpeId                 :: Text
  , cpeProvider           :: Text
  , cpeEnvironment        :: Text
  , cpeProviderEventId    :: Text
  , cpeEventType          :: Text
  , cpeProviderResourceId :: Maybe Text
  , cpeStatus             :: Text
  , cpeAttemptCount       :: Int
  , cpeCheckoutId         :: Maybe Text
  , cpePaymentAttemptId   :: Maybe Text
  , cpeRefundId           :: Maybe Text
  , cpeReceivedAt         :: UTCTime
  , cpeProviderCreatedAt  :: Maybe UTCTime
  , cpeProcessingStartedAt :: Maybe UTCTime
  , cpeLastAttemptAt      :: Maybe UTCTime
  , cpeNextAttemptAt      :: Maybe UTCTime
  , cpeProcessedAt        :: Maybe UTCTime
  , cpeErrorSummary       :: Maybe Text
  } deriving (Eq, Show, Generic)

instance ToJSON CommerceProviderEventDTO
instance FromJSON CommerceProviderEventDTO

data CommerceProviderEventReplayCreate = CommerceProviderEventReplayCreate
  { cperReason :: Text
  } deriving (Eq, Show, Generic)

instance ToJSON CommerceProviderEventReplayCreate
instance FromJSON CommerceProviderEventReplayCreate where
  parseJSON = genericParseJSON strictObjectOptions

type CommerceOperationsAPI =
       "admin" :> "commerce" :> "overview"
         :> Get '[JSON] CommercePaymentOverviewDTO
  :<|> "admin" :> "commerce" :> "provider-events"
         :> QueryParam "status" Text
         :> QueryParam "limit" Int
         :> QueryParam "offset" Int
         :> Get '[JSON] [CommerceProviderEventDTO]
  :<|> "admin" :> "commerce" :> "provider-events"
         :> Capture "eventId" Text
         :> "replay"
         :> ReqBody '[JSON] CommerceProviderEventReplayCreate
         :> Post '[JSON] CommerceProviderEventDTO

File: tdf-hq-ui/src/api/commerceOperations.ts
import { get, post } from './client';

export type CommerceProviderEventStatus =
  | 'pending'
  | 'processing'
  | 'processed'
  | 'retry'
  | 'dead_letter'
  | 'ignored';

export interface CommerceProviderEvent {
  cpeId: string;
  cpeProvider: string;
  cpeEnvironment: 'sandbox' | 'production';
  cpeProviderEventId: string;
  cpeEventType: string;
  cpeProviderResourceId?: string | null;
  cpeStatus: CommerceProviderEventStatus;
  cpeAttemptCount: number;
  cpeCheckoutId?: string | null;
  cpePaymentAttemptId?: string | null;
  cpeRefundId?: string | null;
  cpeReceivedAt: string;
  cpeProviderCreatedAt?: string | null;
  cpeProcessingStartedAt?: string | null;
  cpeLastAttemptAt?: string | null;
  cpeNextAttemptAt?: string | null;
  cpeProcessedAt?: string | null;
  cpeErrorSummary?: string | null;
}

export interface CommerceProviderCapability {
  cpcPaymentMethod: string;
  cpcCapability: string;
  cpcVerificationStatus: string;
  cpcVerifiedAt?: string | null;
}

export interface CommerceProviderAccount {
  cpaProvider: string;
  cpaEnvironment: 'sandbox' | 'production';
  cpaStatus: string;
  cpaContractStatus: string;
  cpaCredentialStatus: string;
  cpaSettlementCurrency: string;
  cpaEnabled: boolean;
  cpaFeatureEnabled: boolean;
  cpaVerifiedAt?: string | null;
  cpaDisabledReason?: string | null;
  cpaCapabilities: CommerceProviderCapability[];
}

export interface CommercePaymentIntentSummary {
  cpiStatus: string;
  cpiCurrency: string;
  cpiCount: number;
  cpiAmountMinor: number;
  cpiAuthorizedMinor: number;
  cpiCapturedMinor: number;
  cpiRefundedMinor: number;
}

export interface CommerceAmountComponentSummary {
  cacComponentType: string;
  cacSource: string;
  cacCurrency: string;
  cacCount: number;
  cacAmountMinor: number;
}

export interface CommerceCommissionSummary {
  ccmProvider: string;
  ccmEnvironment: string;
  ccmCurrency: string;
  ccmCount: number;
  ccmBasisAmountMinor: number;
  ccmCommissionMinor: number;
  ccmProviderFeeMinor: number;
  ccmTaxMinor: number;
  ccmSellerNetMinor: number;
}

export interface CommerceRefundSummary {
  crfProvider: string;
  crfEnvironment: string;
  crfStatus: string;
  crfCurrency: string;
  crfCount: number;
  crfAmountMinor: number;
}

export interface CommerceDisputeSummary {
  cdsProvider: string;
  cdsEnvironment: string;
  cdsKind: string;
  cdsStatus: string;
  cdsCurrency: string;
  cdsCount: number;
  cdsAmountMinor: number;
}

export interface CommerceReconciliationSummary {
  crsProvider: string;
  crsEnvironment: string;
  crsStatus: string;
  crsCurrency?: string | null;
  crsCount: number;
  crsExpectedMinor: number;
  crsActualMinor: number;
}

export interface CommerceSettlementSummary {
  cssProvider: string;
  cssEnvironment: string;
  cssStatus: string;
  cssCurrency: string;
  cssCount: number;
  cssGrossMinor: number;
  cssFeeMinor: number;
  cssWithholdingMinor: number;
  cssRefundMinor: number;
  cssChargebackMinor: number;
  cssNetMinor: number;
}

export interface CommerceSellerBalanceSummary {
  csbProvider: string;
  csbEnvironment: string;
  csbAvailability: string;
  csbCurrency: string;
  csbEntryCount: number;
  csbNetAmountMinor: number;
}

export interface CommercePayoutSummary {
  cpsProvider: string;
  cpsEnvironment: string;
  cpsStatus: string;
  cpsCurrency: string;
  cpsCount: number;
  cpsAmountMinor: number;
}

export interface CommercePaymentOverview {
  cpoGeneratedAt: string;
  cpoProviderAccounts: CommerceProviderAccount[];
  cpoPaymentIntents: CommercePaymentIntentSummary[];
  cpoAmountComponents: CommerceAmountComponentSummary[];
  cpoCommissions: CommerceCommissionSummary[];
  cpoRefunds: CommerceRefundSummary[];
  cpoDisputes: CommerceDisputeSummary[];
  cpoReconciliationExceptions: CommerceReconciliationSummary[];
  cpoSettlements: CommerceSettlementSummary[];
  cpoSellerBalances: CommerceSellerBalanceSummary[];
  cpoPayouts: CommercePayoutSummary[];
}

export const CommerceOperations = {
  getPaymentOverview: () => get<CommercePaymentOverview>('/admin/commerce/overview'),

  listProviderEvents: (params?: {
    status?: CommerceProviderEventStatus;
    limit?: number;
    offset?: number;
  }) => {
    const query = new URLSearchParams();
    if (params?.status) query.set('status', params.status);
    if (params?.limit) query.set('limit', String(params.limit));
    if (params?.offset) query.set('offset', String(params.offset));
    const suffix = query.toString();
    return get<CommerceProviderEvent[]>(
      `/admin/commerce/provider-events${suffix ? `?${suffix}` : ''}`,
    );
  },

  replayProviderEvent: (eventId: string, reason: string) =>
    post<CommerceProviderEvent>(
      `/admin/commerce/provider-events/${encodeURIComponent(eventId)}/replay`,
      { cperReason: reason },
    ),
};
